### PR TITLE
feature: add support for Jar files analysis with JarLauncher (decompile then build AST with spoon)

### PR DESCRIPTION
**Pull request by nharrand**
This PR aims at providing support for JAR annalysis. (Related to #675 ).
My typical usecase is to be able to download an artifact's jar from maven central (and optionally its pom to get dependency information) to decompile it  and build a spoon AST.

This code provide a small wrapper named JarLauncher that does just that.

Still lack:
 * (fixing git commits)
 * some proper testing
 * documentation
 * an option to not decompile but instead build only shadow types?
 * support for alternative decompiler?
 * use maven install plugin for cleaner deployement of CFR?

As the decompiler I use ( [CFR](http://www.benf.org/other/cfr/) ) is not avilable on maven central I propose to add its jar in /libs.
WDYT?

### CODE REVIEW

File: src/main/java/spoon/decompiler/CFRDecompiler.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package spoon.decompiler;
+
+import java.io.File;
+import org.benf.cfr.reader.Main;
+
+public class CFRDecompiler implements Decompiler {
+	@Override
+	public void decompile(File jar, File outputDir) {
+		Main.main(new String[]{jar.getAbsolutePath(), "--outputdir", outputDir.getAbsolutePath()});
```
Comment (monperrus) [inline src/main/java/spoon/decompiler/CFRDecompiler.java]: is their a proper Java API?

File: src/main/java/spoon/decompiler/Decompiler.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+	/**
+	 * Sets the output directory for source generated.
+	 *
+	 * @param jar
+	 * 		Path to jar to be analyzed.
+	 * @param outputDir
+	 * 		Path for the output directory for decompilation.
+	 */
+	void decompile(File jar, File outputDir);
```
Comment (monperrus) [inline src/main/java/spoon/decompiler/Decompiler.java]: What about

 	List<SpoonFile> decompile(String jarPath); 
 	List<SpoonResource> decompile(String jarPath); 
 	CtModel decompile(String jarPath);

File: src/main/java/spoon/JarLauncher.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+import spoon.decompiler.Decompiler;
+import spoon.internal.mavenlauncher.InheritanceModel;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
+
+public class JarLauncher extends Launcher {
```
Comment (monperrus) [inline src/main/java/spoon/JarLauncher.java]: @Experimental

File: pom.xml [hunk 4]
```diff
@@ -295,6 +296,13 @@
       <artifactId>maven-invoker</artifactId>
       <version>3.0.1</version>
     </dependency>
+    <dependency>
+      <groupId>org.benf</groupId>
+      <artifactId>cfr</artifactId>
+      <version>${version.cfr}</version>
+      <scope>system</scope>
```
Comment (msteinbeck) [inline pom.xml]: Scope `system` will not package the dependency into the resulting jar. You need to create a local maven repository and add it to git. See: https://sookocheff.com/post/java/local-maven-repository/

Comment (nharrand) [reply pom.xml]: Thanks. I'll do that.

File: src/main/java/spoon/decompiler/Decompiler.java (hunk 2 above)
Comment (nharrand) [reply src/main/java/spoon/decompiler/Decompiler.java]: I'm not convinced by either of this solutions. My plan was to call the decompiler only if the input jar has been modified after last decompilation or if user forces it. I would prefer having directory declaration in JarLauncher than in Decompiler.

File: src/main/java/spoon/decompiler/CFRDecompiler.java (hunk 1 above)
Comment (nharrand) [reply src/main/java/spoon/decompiler/CFRDecompiler.java]: Not as far as I know.

**Comment by nharrand** [comment]
Ready for review.

**Comment by monperrus** [comment]
LGTM.

Could you add the corresponding documentation in `doc`?

File: pom.xml [hunk 5]
```diff
@@ -306,6 +306,9 @@
           <groupId>org.benf</groupId>
           <artifactId>cfr</artifactId>
           <version>0.132.0</version>
+          <!-- since it is not on Maven central, we don't package it -->
+          <!-- if you need it, download it at http://www.benf.org/other/cfr/ and put the jar in the classpath yourself -->
+          <scope>provided</scope>
```
Comment (nharrand) [inline pom.xml]: It is not on maven central, but I've upload it on `maven.inria.fr`, so it is still publicly available. (So we could package it if we wanted it.)

**Comment by monperrus** [comment]
What's the license of CFR? Is it compatible with LGPL?

**Comment by nharrand** [comment]
It's MIT.

**Pull request merged**